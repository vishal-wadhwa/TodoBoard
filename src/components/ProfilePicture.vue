<template>
  <!--
  Simple wrapper component for vue file pond with all needed props preset.
  -->
  <file-pond
    image-validate-size-min-width='10'
    image-validate-size-min-height='10'
    label-idle='<span class="filepond--label-action">Upload</span> a profile picture'
    max-file-size='1MB'
    name='profile-pic'
    ref='ppic'
    style-panel-layout='compact circle'
    style-load-indicator-position='center bottom'
    style-progress-indicator-position='right bottom'
    style-button-process-item-position='right bottom'
    style-button-remove-item-position='center bottom'
    :accepted-file-types='["image/jpeg", "image/png"]'
    :files='imgs'
    :image-preview-height='height - 10'
    :image-resize-target-width='height'
    :image-resize-target-height='height'
    :server='serverConfig'
    @error='onError'
    @warning='onWarning'
  />
</template>
<script>
import vueFilePond from 'vue-filepond'

import 'filepond/dist/filepond.min.css'
import 'filepond-plugin-image-preview/dist/filepond-plugin-image-preview.min.css'

import FilePondPluginFileValidateSize from 'filepond-plugin-file-validate-size'
import FilePondPluginFileValidateType from 'filepond-plugin-file-validate-type'
import FilePondPluginImageExifOrientation from 'filepond-plugin-image-exif-orientation'
import FilePondPluginImagePreview from 'filepond-plugin-image-preview'
import FilePondPluginImageResize from 'filepond-plugin-image-resize'
import FilePondPluginImageValidateSize from 'filepond-plugin-image-validate-size'

const FilePond = vueFilePond(
  FilePondPluginFileValidateSize,
  FilePondPluginFileValidateType,
  FilePondPluginImageExifOrientation,
  FilePondPluginImagePreview,
  FilePondPluginImageResize,
  FilePondPluginImageValidateSize)

export default {
  components: {
    FilePond
  },
  props: {
    defaultImageSrc: String,
    height: {
      type: [String, Number],
      default: 200
    },
    serverConfig: Object
  },
  data () {
    return {
      imgs: []
    }
  },
  methods: {
    onError (err, file, opts) {
      file.abortLoad()
      file.abortProcessing()
      this.$refs['ppic'].removeFile(file.id)
      this.$emit('pp:error', `Error ${err.code}: ${err.body}.`)
    },
    onWarning (err, file, opts) {
      file.abortLoad()
      file.abortProcessing()
      this.$refs['ppic'].removeFile(file.id)
      this.$emit('pp:warning', `Error ${err.code}: ${err.body}.`)
    }
  },
  created () {
    if (this.defaultImageSrc) this.imgs.push(this.defaultImageSrc)
  }
}
</script>
<style lang="stylus">
.filepond--drop-label {
  color: #4c4e53;
}

.filepond--label-action {
  text-decoration-color: #babdc0;
}

.filepond--panel-root {
  background-color: #edf0f4;
}

.filepond--root {
  width: 200px;
  margin: 0 auto;
  border: 1px dashed black;
}
</style>
