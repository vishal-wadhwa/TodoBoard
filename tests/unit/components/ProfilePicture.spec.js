import { shallowMount, createLocalVue, mount, config } from '@vue/test-utils'
import ProfilePicture from '@/components/ProfilePicture'
import Vuetify from 'vuetify'
import Vue from 'vue'
Vue.use(Vuetify)

config.logModifiedComponents = false

describe('ProfilePicture.vue', () => {
  // required to vuetify [warn] in snapshot test
  // const app = document.createElement('div')
  // app.setAttribute('data-app', true)
  // document.body.appendChild(app)

  let localVue = null
  const factory = (opts = {}, full = false) => {
    return full ? mount(ProfilePicture, opts) : shallowMount(ProfilePicture, opts)
  }
  beforeEach(() => {
    localVue = createLocalVue()
    // localVue.use(Vuetify)
  })

  const errorConstructor = err => `Error ${err.code}: ${err.body}.`
  const file = { abortLoad: jest.fn(), abortProcessing: jest.fn() }
  const stubs = {
    'file-pond': {
      render (h) { return h('div') },
      methods: {
        removeFile: jest.fn()
      }
    }

  }

  it('has empty imgs data array if no default image is provided', () => {
    const wrapper = factory({
      localVue,
      stubs
    })

    expect(wrapper.vm.imgs).toBeInstanceOf(Array)
    expect(wrapper.vm.imgs.length).toBe(0)
  })

  it('has default image, if provided, loaded into imgs data array', () => {
    const defaultImageSrc = 'random-image'
    const wrapper = factory({
      propsData: { defaultImageSrc },
      localVue,
      stubs
    })

    expect(wrapper.vm.imgs).toBeInstanceOf(Array)
    expect(wrapper.vm.imgs.length).toBe(1)
    expect(wrapper.vm.imgs[0]).toBe(defaultImageSrc)
  })

  it('emits "pp:error"', () => {
    const wrapper = factory({
      localVue,
      stubs
    })

    const err = { body: 'Not found', code: 404 }
    wrapper.vm.onError(err, file)

    const emitArr = wrapper.emitted('pp:error')

    expect(emitArr.length).toBe(1)
    expect(emitArr[0].length).toBe(1)
    expect(emitArr[0][0]).toBe(errorConstructor(err))
  })

  it('emits "pp:warning"', () => {
    const wrapper = factory({
      localVue,
      stubs
    })

    const err = { body: 'Can\'t load', code: 400 }
    wrapper.vm.onWarning(err, file)

    const emitArr = wrapper.emitted('pp:warning')

    expect(emitArr.length).toBe(1)
    expect(emitArr[0].length).toBe(1)
    expect(emitArr[0][0]).toBe(errorConstructor(err))
  })

  it('should match snapshot', () => {
    const wrapper = factory({
      localVue,
      stubs
    }, true)

    expect(wrapper.html()).toMatchSnapshot()
  })
})
