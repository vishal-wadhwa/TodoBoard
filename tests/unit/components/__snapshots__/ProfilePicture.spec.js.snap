// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProfilePicture.vue should match snapshot 1`] = `<div image-validate-size-min-width="10" image-validate-size-min-height="10" label-idle="<span class=&quot;filepond--label-action&quot;>Upload</span> a profile picture" max-file-size="1MB" name="profile-pic" style-panel-layout="compact circle" style-load-indicator-position="center bottom" style-progress-indicator-position="right bottom" style-button-process-item-position="right bottom" style-button-remove-item-position="center bottom" accepted-file-types="image/jpeg,image/png" files="" image-preview-height="190" image-resize-target-width="200" image-resize-target-height="200"></div>`;
